### ISSUE TITLE
Support reading annotated variables

### ISSUE THREAD

**Issue by bswck** [jaraco/pip-run#107]
Annotating `__requires__`/`__index_url__` might be discouraged because `pip-run` only supports literal assignments of `__requires__`/`__index_url__`.

However, since `__requires__` is expected to be a immutable literal defined in one place, I'd suggest:
- supporting reading annotated variables because of `typing.Final` which signals: "hey, you shouldn't be reassigning this name `__requires__` here"
- recommending to use immutable type/annotating with a immutable type, e.g. `__requires__: Final[Sequence[str]]` or `__requires__: Final[tuple[str, ...]]`.

This practice has a bunch of users. Let's check for `__all__`, which is a common package-related module-level dunder name, typically not expected to change dynamically:
https://github.com/search?q=%22__all__%3A+list%22&type=code
https://github.com/search?q=%22__all__%3A+tuple%22&type=code
https://github.com/search?q=%22__all__%3A+Final%22&type=code

This feature (1) is easy to support (it's basically just one additional AST node with even simpler interface (2) increases the compatibility with all Python codebases.

I do think this feature is in `pip-run`'s philosophy and makes its usability wider at not much of a cost.

**Comment by jaraco** [comment]
Nice. Thanks for documenting the motivation and especially finding prior art. I'd not thought to look at `__all__`. I particularly appreciate the ability to declare the variable `Final`, which apparently [can be done](https://github.com/RomainBrault/python-whiteprint/blob/58af18f328cc17c27533625aef05bb0c1f449cdd/src/python_whiteprint/loc.py#L12) without overspecifying the type.

I do see the value in supporting valid syntax. Especially since this library is being shared for more than just the `pip-run` cli.

On the other hand, there's lots of valid Python syntax that this construct doesn't support, such as mutation (`__requires__.append('another')`) or multi-value assignment (`__requires__, __index_url__ = ...`) or non-literal assignment (`__requires__ = [...] + [...]`).

I'm a little worried that nobody has asked for this feature. Are you wanting for yourself to add type annotations to a `__requires__` directive? Do you know of any other users who would use this feature if it existed?

I'm also apprehensive as it starts to violate the zen of "preferably one obvious way to do it".

Overall, I'm inclined to say we should do it, especially if the functional approach can be maintained (such as drafted in c9a7094).

**Comment by bswck** [comment]
> I'm also apprehensive as it starts to violate the zen of "preferably one obvious way to do it".

But does it?
Depends on how you describe it.

I do believe that we could say there is no more ways to do it with this feature: the obvious way to specify requirements without passing them via `sys.argv` to pip-run (which already means there are two ways) is to declare a sequence of strings (or a single string! doesn't it already violate the zen? there are three ways already) storing the dependencies.

Syntax isn't what violates the Zen here, I believe. I think it's just about some basic flexibility, but that's an opinion, not a helpful fact.

**Comment by bswck** [comment]
> On the other hand, there's lots of valid Python syntax that this construct doesn't support, such as mutation (`__requires__.append('another')`) or multi-value assignment (`__requires__, __index_url__ = ...`) or non-literal assignment (`__requires__ = [...] + [...]`).

Yes. Additionally, we do not support `(__requires__ := ...)` which does set the value at runtime.
But these things were obvious since the beginning.
If Python had something called "build time", we could be more dynamic and actually collect these variables from the "build time", instead of analyzing the code.

> I'm a little worried that nobody has asked for this feature. Are you wanting for yourself to add type annotations to a __requires__ directive? Do you know of any other users who would use this feature if it existed?

Well, let me be the first one then. I wanted to declare a requirement list _without overspecifying the type_, that is `__requires__: Final = [...]` :)

And I'm pretty sure some people would use it. But I can't speak for them.

**Comment by bswck** [comment]
> I promise to be more forgiving for another PR.

Alright, see you there!

**Comment by jaraco** [comment]
> Consider using [c9a7094](https://github.com/jaraco/pip-run/commit/c9a7094f83328853f58e562cb253f5d5fb8a6699) as a baseline (though not required).

I've merged that change, so now it's required ;)

Let me know if you're still interested in working on this; otherwise, I'll unassign it. No obligation.
